File: LeetCode/Medium/Q875.java
package LeetCode.Medium;

public class Q875 {
    public int minEatingSpeed(int[] piles, int h) {
        int s = 1 ; int e = Arrays.stream(piles).max().getAsInt();
        int ans = e;
        while(s <= e){
            int m = s + ( e - s) / 2;    
            if(canFinish(piles,h ,m)){
                ans = m;
                e = m - 1;
            }
            else{
                s = m + 1;
            }                        
        }
        return ans;

    }
    public boolean canFinish(int[] piles , int h , int k){
        long hours = 0;
        for(int pile: piles){
            hours+=pile/k;
            if(pile% k!=0) hours ++;
        }
        return hours <=h;
    }
}

File: LeetCode/Python/Medium/Q875.py
class Solution:
    def minEatingSpeed(self, piles, h) -> int:
        def canFinish(k: int) -> bool:
            hours = 0
            for pile in piles:
                hours += pile // k
                if pile % k != 0:
                    hours += 1
                if hours > h:
                    return False
            return True

        start, end = 1, max(piles)
        ans = end
        while start <= end:
            mid = start + (end - start) // 2
            if canFinish(mid):
                ans = mid
                end = mid - 1
            else:
                start = mid + 1
        return ans
